// @livekit/rtc-node/dist/audio_stream.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var audio_stream_exports = {};
__export(audio_stream_exports, {
  AudioStream: () => AudioStream,
  AudioStreamSource: () => AudioStreamSource
});
module.exports = __toCommonJS(audio_stream_exports);
var import_rtc_ffi_bindings = require("@livekit/rtc-ffi-bindings");
var import_audio_frame = require("./audio_frame.cjs");
var import_ffi_client = require("./ffi_client.cjs");
var import_frame_processor = require("./frame_processor.cjs");
var import_log = require("./log.cjs");
class AudioStreamSource {
  constructor(track, sampleRateOrOptions, numChannels) {
    this.disposed = false;
    this.frameProcessor = null;
    this.autoCloseProcessor = true;
    this.onEvent = (ev) => {
      if (!this.controller) {
        throw new Error("Stream controller not initialized");
      }
      if (ev.message.case != "audioStreamEvent" || ev.message.value.streamHandle != this.ffiHandle.handle) {
        return;
      }
      const streamEvent = ev.message.value.message;
      switch (streamEvent.case) {
        case "frameReceived":
          let frame = import_audio_frame.AudioFrame.fromOwnedInfo(streamEvent.value.frame);
          if (this.frameProcessor && this.frameProcessor.isEnabled()) {
            try {
              frame = this.frameProcessor.process(frame);
            } catch (err) {
              import_log.log.warn(`Frame processing failed, passing through original frame: ${err}`);
            }
          }
          this.controller.enqueue(frame);
          break;
        case "eos":
          import_ffi_client.FfiClient.instance.off(import_ffi_client.FfiClientEvent.FfiEvent, this.onEvent);
          this.controller.close();
          if (!this.disposed) {
            this.disposed = true;
            this.track.unregisterAudioStream(this);
            this.ffiHandle.dispose();
            if (this.frameProcessor && this.autoCloseProcessor) {
              this.frameProcessor.close();
            }
          }
          break;
      }
    };
    this.track = track;
    if (sampleRateOrOptions !== void 0 && typeof sampleRateOrOptions !== "number") {
      this.sampleRate = sampleRateOrOptions.sampleRate ?? 48e3;
      this.numChannels = sampleRateOrOptions.numChannels ?? 1;
      if ((0, import_frame_processor.isFrameProcessor)(sampleRateOrOptions.noiseCancellation)) {
        this.frameProcessor = sampleRateOrOptions.noiseCancellation;
        this.autoCloseProcessor = sampleRateOrOptions.autoCloseNoiseCancellation ?? true;
      } else {
        this.legacyNcOptions = sampleRateOrOptions.noiseCancellation;
      }
      this.frameSizeMs = sampleRateOrOptions.frameSizeMs;
    } else {
      this.sampleRate = sampleRateOrOptions ?? 48e3;
      this.numChannels = numChannels ?? 1;
    }
    const req = new import_rtc_ffi_bindings.NewAudioStreamRequest({
      type: import_rtc_ffi_bindings.AudioStreamType.AUDIO_STREAM_NATIVE,
      trackHandle: track.ffi_handle.handle,
      sampleRate: this.sampleRate,
      numChannels: this.numChannels,
      frameSizeMs: this.frameSizeMs,
      ...this.legacyNcOptions ? {
        audioFilterModuleId: this.legacyNcOptions.moduleId,
        audioFilterOptions: JSON.stringify(this.legacyNcOptions.options)
      } : {}
    });
    const res = import_ffi_client.FfiClient.instance.request({
      message: {
        case: "newAudioStream",
        value: req
      }
    });
    this.ffiHandle = new import_ffi_client.FfiHandle(res.stream.handle.id);
    import_ffi_client.FfiClient.instance.on(import_ffi_client.FfiClientEvent.FfiEvent, this.onEvent);
    track.registerAudioStream(this);
  }
  /** @internal */
  get processor() {
    return this.frameProcessor;
  }
  start(controller) {
    this.controller = controller;
  }
  cancel() {
    import_ffi_client.FfiClient.instance.off(import_ffi_client.FfiClientEvent.FfiEvent, this.onEvent);
    if (!this.disposed) {
      this.disposed = true;
      this.track.unregisterAudioStream(this);
      this.ffiHandle.dispose();
      if (this.frameProcessor && this.autoCloseProcessor) {
        this.frameProcessor.close();
      }
    }
  }
}
class AudioStream extends ReadableStream {
  constructor(track, sampleRateOrOptions, numChannels) {
    super(new AudioStreamSource(track, sampleRateOrOptions, numChannels));
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AudioStream,
  AudioStreamSource
});
