// @livekit/rtc-node/dist/ffi_client.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ffi_client_exports = {};
__export(ffi_client_exports, {
  FfiClient: () => FfiClient,
  FfiClientEvent: () => FfiClientEvent,
  FfiHandle: () => import_rtc_ffi_bindings.FfiHandle,
  FfiRequest: () => import_rtc_ffi_bindings.FfiRequest,
  dispose: () => import_rtc_ffi_bindings.livekitDispose
});
module.exports = __toCommonJS(ffi_client_exports);
var import_rtc_ffi_bindings = require("@livekit/rtc-ffi-bindings");
var import_events = __toESM(require("events"), 1);
var import_version = require("./version.cjs");
var FfiClientEvent = /* @__PURE__ */ ((FfiClientEvent2) => {
  FfiClientEvent2["FfiEvent"] = "ffi_event";
  return FfiClientEvent2;
})(FfiClientEvent || {});
class FfiClient extends import_events.default {
  constructor() {
    super();
    this._nextRequestAsyncId = BigInt(0);
    this.setMaxListeners(0);
    (0, import_rtc_ffi_bindings.livekitInitialize)(
      (event_data) => {
        const event = import_rtc_ffi_bindings.FfiEvent.fromBinary(event_data);
        this.emit("ffi_event" /* FfiEvent */, event);
      },
      true,
      import_version.SDK_VERSION
    );
  }
  /** @internal */
  static get instance() {
    if (!globalThis._ffiClientInstance) {
      globalThis._ffiClientInstance = new FfiClient();
    }
    return globalThis._ffiClientInstance;
  }
  request(req) {
    const request = new import_rtc_ffi_bindings.FfiRequest(req);
    const req_data = request.toBinary();
    const res_data = (0, import_rtc_ffi_bindings.livekitFfiRequest)(req_data);
    return import_rtc_ffi_bindings.FfiResponse.fromBinary(res_data).message.value;
  }
  copyBuffer(ptr, len) {
    return (0, import_rtc_ffi_bindings.livekitCopyBuffer)(ptr, len);
  }
  retrievePtr(data) {
    return (0, import_rtc_ffi_bindings.livekitRetrievePtr)(data);
  }
  getNextRequestAsyncId() {
    return ++this._nextRequestAsyncId;
  }
  async waitFor(predicate, options) {
    return new Promise((resolve, reject) => {
      var _a, _b;
      const listener = (ev) => {
        if (predicate(ev)) {
          cleanup();
          resolve(ev.message.value);
        }
      };
      const cleanup = () => {
        var _a2;
        this.off("ffi_event" /* FfiEvent */, listener);
        (_a2 = options == null ? void 0 : options.signal) == null ? void 0 : _a2.removeEventListener("abort", onAbort);
      };
      const onAbort = () => {
        var _a2;
        cleanup();
        reject(((_a2 = options == null ? void 0 : options.signal) == null ? void 0 : _a2.reason) ?? new Error("waitFor aborted"));
      };
      if ((_a = options == null ? void 0 : options.signal) == null ? void 0 : _a.aborted) {
        reject(options.signal.reason ?? new Error("waitFor aborted"));
        return;
      }
      (_b = options == null ? void 0 : options.signal) == null ? void 0 : _b.addEventListener("abort", onAbort);
      this.on("ffi_event" /* FfiEvent */, listener);
    });
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FfiClient,
  FfiClientEvent,
  FfiHandle,
  FfiRequest,
  dispose
});
