// @livekit/rtc-node/dist/ffi_client.js
import {
  FfiEvent,
  FfiHandle,
  FfiRequest,
  FfiResponse,
  livekitCopyBuffer,
  livekitDispose,
  livekitFfiRequest,
  livekitInitialize,
  livekitRetrievePtr
} from "@livekit/rtc-ffi-bindings";
import EventEmitter from "events";
import { SDK_VERSION } from "./version.js";
var FfiClientEvent = /* @__PURE__ */ ((FfiClientEvent2) => {
  FfiClientEvent2["FfiEvent"] = "ffi_event";
  return FfiClientEvent2;
})(FfiClientEvent || {});
class FfiClient extends EventEmitter {
  constructor() {
    super();
    this._nextRequestAsyncId = BigInt(0);
    this.setMaxListeners(0);
    livekitInitialize(
      (event_data) => {
        const event = FfiEvent.fromBinary(event_data);
        this.emit("ffi_event" /* FfiEvent */, event);
      },
      true,
      SDK_VERSION
    );
  }
  /** @internal */
  static get instance() {
    if (!globalThis._ffiClientInstance) {
      globalThis._ffiClientInstance = new FfiClient();
    }
    return globalThis._ffiClientInstance;
  }
  request(req) {
    const request = new FfiRequest(req);
    const req_data = request.toBinary();
    const res_data = livekitFfiRequest(req_data);
    return FfiResponse.fromBinary(res_data).message.value;
  }
  copyBuffer(ptr, len) {
    return livekitCopyBuffer(ptr, len);
  }
  retrievePtr(data) {
    return livekitRetrievePtr(data);
  }
  getNextRequestAsyncId() {
    return ++this._nextRequestAsyncId;
  }
  async waitFor(predicate, options) {
    return new Promise((resolve, reject) => {
      var _a, _b;
      const listener = (ev) => {
        if (predicate(ev)) {
          cleanup();
          resolve(ev.message.value);
        }
      };
      const cleanup = () => {
        var _a2;
        this.off("ffi_event" /* FfiEvent */, listener);
        (_a2 = options == null ? void 0 : options.signal) == null ? void 0 : _a2.removeEventListener("abort", onAbort);
      };
      const onAbort = () => {
        var _a2;
        cleanup();
        reject(((_a2 = options == null ? void 0 : options.signal) == null ? void 0 : _a2.reason) ?? new Error("waitFor aborted"));
      };
      if ((_a = options == null ? void 0 : options.signal) == null ? void 0 : _a.aborted) {
        reject(options.signal.reason ?? new Error("waitFor aborted"));
        return;
      }
      (_b = options == null ? void 0 : options.signal) == null ? void 0 : _b.addEventListener("abort", onAbort);
      this.on("ffi_event" /* FfiEvent */, listener);
    });
  }
}
export {
  FfiClient,
  FfiClientEvent,
  FfiHandle,
  FfiRequest,
  livekitDispose as dispose
};
