// @livekit/rtc-node/dist/frame_processor.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var frame_processor_exports = {};
__export(frame_processor_exports, {
  FrameProcessor: () => FrameProcessor,
  FrameProcessorSymbol: () => FrameProcessorSymbol,
  isFrameProcessor: () => isFrameProcessor
});
module.exports = __toCommonJS(frame_processor_exports);
var import_audio_frame = require("./audio_frame.cjs");
var import_video_frame = require("./video_frame.cjs");
const FrameProcessorSymbol = Symbol.for("lk.frame-processor");
function isFrameProcessor(maybeProcessor) {
  return maybeProcessor !== null && typeof maybeProcessor === "object" && "symbol" in maybeProcessor && maybeProcessor.symbol === FrameProcessorSymbol;
}
class FrameProcessor {
  constructor() {
    this.symbol = FrameProcessorSymbol;
  }
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  onStreamInfoUpdated(_info) {
  }
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  onStreamInfoCleared() {
  }
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  onCredentialsUpdated(_credentials) {
  }
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  onCredentialsCleared() {
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FrameProcessor,
  FrameProcessorSymbol,
  isFrameProcessor
});
