// @livekit/rtc-node/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  AudioFilter: () => import_audio_filter.AudioFilter,
  AudioFrame: () => import_audio_frame.AudioFrame,
  AudioMixer: () => import_audio_mixer.AudioMixer,
  AudioResampler: () => import_audio_resampler.AudioResampler,
  AudioResamplerQuality: () => import_audio_resampler.AudioResamplerQuality,
  AudioSource: () => import_audio_source.AudioSource,
  AudioStream: () => import_audio_stream.AudioStream,
  ConnectError: () => import_room.ConnectError,
  ConnectionQuality: () => import_rtc_ffi_bindings3.ConnectionQuality,
  ConnectionState: () => import_rtc_ffi_bindings3.ConnectionState,
  ContinualGatheringPolicy: () => import_rtc_ffi_bindings3.ContinualGatheringPolicy,
  DataPacketKind: () => import_rtc_ffi_bindings3.DataPacketKind,
  DisconnectReason: () => import_rtc_ffi_bindings2.DisconnectReason,
  E2EEManager: () => import_e2ee.E2EEManager,
  EncryptionState: () => import_rtc_ffi_bindings.EncryptionState,
  EncryptionType: () => import_rtc_ffi_bindings.EncryptionType,
  FrameCryptor: () => import_e2ee.FrameCryptor,
  FrameProcessor: () => import_frame_processor.FrameProcessor,
  IceServer: () => import_rtc_ffi_bindings3.IceServer,
  IceTransportType: () => import_rtc_ffi_bindings3.IceTransportType,
  KeyProvider: () => import_e2ee.KeyProvider,
  LocalAudioTrack: () => import_track.LocalAudioTrack,
  LocalParticipant: () => import_participant.LocalParticipant,
  LocalTrackPublication: () => import_track_publication.LocalTrackPublication,
  LocalVideoTrack: () => import_track.LocalVideoTrack,
  Participant: () => import_participant.Participant,
  ParticipantKind: () => import_rtc_ffi_bindings2.ParticipantKind,
  ParticipantKindDetail: () => import_rtc_ffi_bindings2.ParticipantKindDetail,
  RemoteAudioTrack: () => import_track.RemoteAudioTrack,
  RemoteParticipant: () => import_participant.RemoteParticipant,
  RemoteTrackPublication: () => import_track_publication.RemoteTrackPublication,
  RemoteVideoTrack: () => import_track.RemoteVideoTrack,
  Room: () => import_room.Room,
  RoomEvent: () => import_room.RoomEvent,
  RpcError: () => import_rpc.RpcError,
  SimulateScenarioKind: () => import_rtc_ffi_bindings4.SimulateScenarioKind,
  StreamState: () => import_rtc_ffi_bindings4.StreamState,
  Track: () => import_track.Track,
  TrackKind: () => import_rtc_ffi_bindings4.TrackKind,
  TrackPublication: () => import_track_publication.TrackPublication,
  TrackPublishOptions: () => import_rtc_ffi_bindings3.TrackPublishOptions,
  TrackSource: () => import_rtc_ffi_bindings4.TrackSource,
  VideoBufferType: () => import_rtc_ffi_bindings5.VideoBufferType,
  VideoCodec: () => import_rtc_ffi_bindings5.VideoCodec,
  VideoFrame: () => import_video_frame.VideoFrame,
  VideoRotation: () => import_rtc_ffi_bindings5.VideoRotation,
  VideoSource: () => import_video_source.VideoSource,
  VideoStream: () => import_video_stream.VideoStream,
  combineAudioFrames: () => import_audio_frame.combineAudioFrames,
  dispose: () => import_ffi_client.dispose,
  isFrameProcessor: () => import_frame_processor.isFrameProcessor
});
module.exports = __toCommonJS(index_exports);
var import_audio_frame = require("./audio_frame.cjs");
var import_audio_resampler = require("./audio_resampler.cjs");
var import_audio_source = require("./audio_source.cjs");
var import_audio_stream = require("./audio_stream.cjs");
var import_audio_filter = require("./audio_filter.cjs");
var import_audio_mixer = require("./audio_mixer.cjs");
__reExport(index_exports, require("./data_streams/index.cjs"), module.exports);
var import_e2ee = require("./e2ee.cjs");
var import_ffi_client = require("./ffi_client.cjs");
var import_participant = require("./participant.cjs");
var import_rtc_ffi_bindings = require("@livekit/rtc-ffi-bindings");
var import_rtc_ffi_bindings2 = require("@livekit/rtc-ffi-bindings");
var import_rtc_ffi_bindings3 = require("@livekit/rtc-ffi-bindings");
var import_rtc_ffi_bindings4 = require("@livekit/rtc-ffi-bindings");
var import_rtc_ffi_bindings5 = require("@livekit/rtc-ffi-bindings");
var import_room = require("./room.cjs");
var import_rpc = require("./rpc.cjs");
var import_track = require("./track.cjs");
var import_track_publication = require("./track_publication.cjs");
var import_video_frame = require("./video_frame.cjs");
var import_video_source = require("./video_source.cjs");
var import_video_stream = require("./video_stream.cjs");
var import_frame_processor = require("./frame_processor.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AudioFilter,
  AudioFrame,
  AudioMixer,
  AudioResampler,
  AudioResamplerQuality,
  AudioSource,
  AudioStream,
  ConnectError,
  ConnectionQuality,
  ConnectionState,
  ContinualGatheringPolicy,
  DataPacketKind,
  DisconnectReason,
  E2EEManager,
  EncryptionState,
  EncryptionType,
  FrameCryptor,
  FrameProcessor,
  IceServer,
  IceTransportType,
  KeyProvider,
  LocalAudioTrack,
  LocalParticipant,
  LocalTrackPublication,
  LocalVideoTrack,
  Participant,
  ParticipantKind,
  ParticipantKindDetail,
  RemoteAudioTrack,
  RemoteParticipant,
  RemoteTrackPublication,
  RemoteVideoTrack,
  Room,
  RoomEvent,
  RpcError,
  SimulateScenarioKind,
  StreamState,
  Track,
  TrackKind,
  TrackPublication,
  TrackPublishOptions,
  TrackSource,
  VideoBufferType,
  VideoCodec,
  VideoFrame,
  VideoRotation,
  VideoSource,
  VideoStream,
  combineAudioFrames,
  dispose,
  isFrameProcessor,
  ...require("./data_streams/index.cjs")
});
