// @livekit/rtc-node/dist/track.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var track_exports = {};
__export(track_exports, {
  LocalAudioTrack: () => LocalAudioTrack,
  LocalVideoTrack: () => LocalVideoTrack,
  RemoteAudioTrack: () => RemoteAudioTrack,
  RemoteVideoTrack: () => RemoteVideoTrack,
  Track: () => Track
});
module.exports = __toCommonJS(track_exports);
var import_rtc_ffi_bindings = require("@livekit/rtc-ffi-bindings");
var import_ffi_client = require("./ffi_client.cjs");
class Track {
  constructor(owned) {
    this.roomRef = null;
    this.audioStreams = /* @__PURE__ */ new Set();
    this.onRoomTokenRefreshed = () => {
      const room = this.resolveRoom();
      if (!room || !room.token || !room.serverUrl) {
        return;
      }
      for (const stream of this.iterateStreams()) {
        const processor = stream.processor;
        if (!processor) {
          continue;
        }
        processor.onCredentialsUpdated({ token: room.token, url: room.serverUrl });
      }
    };
    this.info = owned.info;
    this.ffi_handle = new import_ffi_client.FfiHandle(owned.handle.id);
    this.streamFinalizationRegistry = new FinalizationRegistry(
      (ref) => {
        this.audioStreams.delete(ref);
      }
    );
  }
  /** @internal */
  resolveRoom() {
    var _a;
    return ((_a = this.roomRef) == null ? void 0 : _a.deref()) ?? null;
  }
  /** @internal */
  setRoom(room) {
    const oldRoom = this.resolveRoom();
    if (!oldRoom && !room) {
      return;
    }
    if (oldRoom && oldRoom !== room) {
      oldRoom.off("tokenRefreshed", this.onRoomTokenRefreshed);
    }
    if (room) {
      room.off("tokenRefreshed", this.onRoomTokenRefreshed);
      room.on("tokenRefreshed", this.onRoomTokenRefreshed);
    }
    this.roomRef = room ? new WeakRef(room) : null;
    for (const stream of this.iterateStreams()) {
      this.pushProcessorMetadataToStream(stream, room);
    }
  }
  /** @internal */
  registerAudioStream(stream) {
    const ref = new WeakRef(stream);
    this.audioStreams.add(ref);
    this.streamFinalizationRegistry.register(stream, ref);
    const room = this.resolveRoom();
    if (room) {
      this.pushProcessorMetadataToStream(stream, room);
    }
  }
  /** @internal */
  unregisterAudioStream(stream) {
    for (const ref of this.audioStreams) {
      if (ref.deref() === stream) {
        this.audioStreams.delete(ref);
        return;
      }
    }
  }
  *iterateStreams() {
    const dead = [];
    for (const ref of this.audioStreams) {
      const stream = ref.deref();
      if (stream) {
        yield stream;
      } else {
        dead.push(ref);
      }
    }
    for (const ref of dead) {
      this.audioStreams.delete(ref);
    }
  }
  pushProcessorMetadataToStream(stream, room) {
    var _a, _b;
    const processor = stream.processor;
    if (!processor) {
      return;
    }
    if (!room) {
      (_a = processor.onStreamInfoCleared) == null ? void 0 : _a.call(processor);
      (_b = processor.onCredentialsCleared) == null ? void 0 : _b.call(processor);
      return;
    }
    let identity = "";
    let publicationSid = "";
    const trackSid = this.sid;
    if (trackSid) {
      let found = false;
      for (const participant of room.remoteParticipants.values()) {
        const publication = participant.trackPublications.get(trackSid);
        if (publication) {
          identity = participant.identity;
          publicationSid = publication.sid ?? "";
          found = true;
          break;
        }
      }
      if (!found) {
        const local = room.localParticipant;
        if (local) {
          for (const publication of local.trackPublications.values()) {
            if (publication.sid === trackSid) {
              identity = local.identity;
              publicationSid = publication.sid ?? "";
              break;
            }
          }
        }
      }
    }
    processor.onStreamInfoUpdated({
      roomName: room.name ?? "",
      participantIdentity: identity,
      publicationSid
    });
    if (room.token && room.serverUrl) {
      processor.onCredentialsUpdated({ token: room.token, url: room.serverUrl });
    }
  }
  get sid() {
    var _a;
    return (_a = this.info) == null ? void 0 : _a.sid;
  }
  get name() {
    var _a;
    return (_a = this.info) == null ? void 0 : _a.name;
  }
  get kind() {
    var _a;
    return (_a = this.info) == null ? void 0 : _a.kind;
  }
  get stream_state() {
    var _a;
    return (_a = this.info) == null ? void 0 : _a.streamState;
  }
  get muted() {
    var _a;
    return (_a = this.info) == null ? void 0 : _a.muted;
  }
  async close() {
    this.ffi_handle.dispose();
  }
}
class LocalAudioTrack extends Track {
  constructor(owned, source) {
    super(owned);
    this.source = source;
  }
  static createAudioTrack(name, source) {
    const req = new import_rtc_ffi_bindings.CreateAudioTrackRequest({
      name,
      sourceHandle: source.ffiHandle.handle
    });
    const res = import_ffi_client.FfiClient.instance.request({
      message: { case: "createAudioTrack", value: req }
    });
    return new LocalAudioTrack(res.track, source);
  }
  async close(closeSource = true) {
    var _a;
    await super.close();
    if (closeSource) {
      await ((_a = this.source) == null ? void 0 : _a.close());
    }
  }
}
class LocalVideoTrack extends Track {
  constructor(owned, source) {
    super(owned);
    this.source = source;
  }
  static createVideoTrack(name, source) {
    const req = new import_rtc_ffi_bindings.CreateVideoTrackRequest({
      name,
      sourceHandle: source.ffiHandle.handle
    });
    const res = import_ffi_client.FfiClient.instance.request({
      message: { case: "createVideoTrack", value: req }
    });
    return new LocalVideoTrack(res.track, source);
  }
  async close(closeSource = true) {
    var _a;
    await super.close();
    if (closeSource) {
      await ((_a = this.source) == null ? void 0 : _a.close());
    }
  }
}
class RemoteVideoTrack extends Track {
  constructor(owned) {
    super(owned);
  }
}
class RemoteAudioTrack extends Track {
  constructor(owned) {
    super(owned);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LocalAudioTrack,
  LocalVideoTrack,
  RemoteAudioTrack,
  RemoteVideoTrack,
  Track
});
